File: medium/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        /*
        leftNode.val < root.value
        rightNode.val > root.value
        
        1<=Number of Nodes <=10^4
        each Node's value => Integer range
        
        TC : O(n)
        SC : O(n)
        */
        
        return isValidTree(root, Long.MAX_VALUE, Long.MIN_VALUE);
    }
    
    private boolean isValidTree(TreeNode node, long max, long min){
        if(node == null) return true;
        if(node.val >= max || node.val <= min) return false;
        
        return isValidTree(node.left, node.val, min) && isValidTree(node.right, max, node.val);
    }
}

File: medium/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        '''
        - check left node  :
        if the maximun valus is lower than root.val & left tree is valid => left is true
        - check right node :
        if the minimun valus is bigger than root.val & right tree is valid => right is true
        - if left and right is valid => current tree is valid!
        => Recursive
        TC : O(n)
        SC : O(n)
        '''
        def helper(max_node, min_node, root) :
            if root == None : return True
            if root.val >= max_node or root.val <= min_node : return False
            
            if helper(max_node, root.val, root.right) and helper(root.val, min_node, root.left) :
                return True
        
        right = helper(math.inf, root.val, root.right)
        left = helper(root.val , -math.inf, root.left)
        
        return right and left
        
    
        
        
